meeteval/viz/visualize.js

```diff
@@ -1489,6 +1491,45 @@ class CanvasPlot {
         }
     }
 
+    function drawRecordingAudioButton(container, rangeSelector, name, recording_file) {
+        let update_audio = function (){
+            // Note: Deleting and adding an audio with the same content doesn't
+            //       trigger a load. Hence, no optimization necessary.
+            audio_div.selectAll("*").remove();
+            let lower = rangeSelector.lower_input.node().value;
+            let upper = rangeSelector.upper_input.node().value;
+            let value = "[" + lower + ":" + upper + "]";
+            let path = server_path.node().value + "/" + file_path.node().value + "::" + value;
+            if ( parseFloat(lower) < parseFloat(upper) ){
+                audio_div.append("div").text(value);
+                audio_div.append("audio")
+                    .on("error", () => {audio_div.append("div").text("Issue while loading the audio")})
+                    .attr("controls", "true")
+                    .attr("src", path).text(path);
+            } else {
+                audio_div.append("div").text("Invalid range: " + lower + " - " + upper);
+            }
+        };
+
+        let pill = container.append("div").classed("pill", true);
+        let audio_tooltip = addTooltip(pill);
+
+        pill.append("span").text((name.trim() !== "") ? "🔊 " + name : "🔊");
+
+        let input = audio_tooltip.append("div").style("display", "flex");
+        let server_path = input.append("input").attr("type", "text").attr("placeholder", "e.g. http://localhost:7777").property("value", "http://localhost:7777");
+        let file_path = input.append("input").attr("type", "text").style("width", "30em").attr("placeholder", "e.g. /path/to/file.wav").property("value", recording_file);
+
+        let audio_div = audio_tooltip.append("div").style("display", "flex");
+
+        input.append("button").text("Load").on("click", () => update_audio());
+        server_path.on("keydown", (event) => {if (event.key === "Enter") {update_audio();}});
+        file_path.on("keydown", (event) => {if (event.key === "Enter") {update_audio();}});
+        pill.on("click", () => {update_audio(); audio_div.select("audio").node().play()});
+
```

Comment:
Changing the selection does not update the audio. This might turn out to be annoying. But it's not that easy to get the update in here. You would have to get the details plot, add a callback to onScroll, setup a timer to throttle server requests (update after a certain delay)...

Reply:
I initially had the same idea. But it turned out, that I always clicked on the "pill".

So I made the pill to a button, that triggers an update and the play.

Alternatively, the audio could be updated, when the tooltip appears (the browser does some caching, so it will be relative cheap.).

Reply:
Updating the audio when the tooltip appears sounds like a good idea

Reply:
Now, where I think about this, it might be better to load the audio only, when it is likely to be used. Just moving the mouse over the pill should not trigger a load.

Reply:
What happens if a user hovers over the pill and starts the audio using the audio controls in the tooltip? The user would then get the old audio. I think starting the audio doesn't trigger a reload

Reply:
I fixed this now:
When moving the mouse over the pill, it will be checked, whether the content is valid. If not, it will be deleted.
In this way, the number of triggered audio loads is as small as possible.

Reply:
Okay, sounds good!